Add a reverse index to Multidictionary

kbongort opened the pull request:
Add a reverse index to Multidictionary to make value removal more efficient in large dictionaries.

In a module with several tens of thousands of definitions in the dependency graph, disappearing nodes from the graph becomes very expensive as removeOccurences(of:) is an O(number_of_keys) operation. This change makes removeOccurrences(of:) O(number_of_keys_containing_the_value), which can be tens of thousands of times faster.

The cost of this change is that insertValue(:forKey:) and removeValue(:forKey:) are approximately twice as expensive, and the dictionary will use about twice as much memory. For small graphs, this difference should be trivial. For my several-tens-of-thousands-of-definitions graph, it takes about four seconds to load rather than two. The driver doesn't seem to use significantly more memory overall. Without this change removeOccurences(of:) is so exceedingly expensive (about 0.5 seconds per call) that the increase in load time is well worth it.

This change could be further optimized so that the `reverseIndex` is only built the first time removeOccurences(of:) is called. Then there would be no negative performance impact at all for many builds. However, the negative impact seems to be quite small, so I am not sure the extra complexity of such an optimization is worthwhile.

kbongort commented:
@swift-ci please test

davidungar commented on Sources/SwiftDriver/IncrementalCompilation/Multidictionary.swift:
```diff
@@ -123,8 +123,9 @@
   /// - Parameter v: The value to remove.
   public mutating func removeOccurrences(of v: Value) {
-    for k in self.dictionary.keys {
+    for k in self.reverseIndex[v, default: []] {
       self.dictionary[k]!.remove(v)
     }
+    self.reverseIndex.removeValue(forKey: v)
```
IIIRC, `removeValue` returns the old value if any. So this function only needs one probe of the `reverseIndex` dictionary.

kbongort replied:
Done

davidungar reviewed (changes requested):
Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency?
The code is fairly short and clear now, but over time it may not be.

kbongort commented:
> Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency? The code is fairly short and clear now, but over time it may not be.

Can I make a counter-offer of unit tests for this data structure? I think tests would be preferable to asserts, since they'd aid future developers and add no run-time cost (if only for development builds).

davidungar commented:
> > Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency? The code is fairly short and clear now, but over time it may not be.
> 
> Can I make a counter-offer of unit tests for this data structure? I think tests would be preferable to asserts, since they'd aid future developers and add no run-time cost (if only for development builds).

Interesting thought, but I'm not ready to accept it at this time. Here's why:
1. Unit tests do not make the invariant explicit in the code
2. I suspect that adding an assert can be done with minimal overhead. Either one Dictionary-of-Sets probe in each direction after modification, or a verify pass at the end of the run. I'd feel better going that route.
3. A pitfall of unit tests--which I've fallen into myself from time to time--is that by testing specific inputs, a unit test can miss cases an assert would catch.

kbongort commented on Sources/SwiftDriver/IncrementalCompilation/Multidictionary.swift:
```diff
@@ -101,6 +99,7 @@
   ///            other values for the given key. Else, returns `false.
   @discardableResult
   public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
+    self.reverseIndex[v, default: []].insert(key)
```
Ack - yes I'm surprised too! Makes for very nice syntax in the tests.

kbongort commented:
> > > Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency? The code is fairly short and clear now, but over time it may not be.
> > 
> > 
> > Can I make a counter-offer of unit tests for this data structure? I think tests would be preferable to asserts, since they'd aid future developers and add no run-time cost (if only for development builds).
> 
> Interesting thought, but I'm not ready to accept it at this time. Here's why:
> 
> 1. Unit tests do not make the invariant explicit in the code
> 2. I suspect that adding an assert can be done with minimal overhead. Either one Dictionary-of-Sets probe in each direction after modification, or a verify pass at the end of the run. I'd feel better going that route.
> 3. A pitfall of unit tests--which I've fallen into myself from time to time--is that by testing specific inputs, a unit test can miss cases an assert would catch.

Okay, I added an assert to check that removeOccurrences(of:) removes all occurrences from self.dictionary. I'm having trouble translating your direction from (2) into code that makes sense. I thought about the following:
```
  public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
    self.reverseIndex[v, default: []].insert(key)
    let inserted = self.dictionary[key, default: []].insert(v).inserted
    assert(self.reverseIndex[v]?.contains(key) == true)
    assert(self.dictionary[key]?.contains(v) == true)
    return inserted
  }
```
But it seems silly to assert that the inserts or removals worked as expected. Let me know what you think.

davidungar commented:
> > > > Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency? The code is fairly short and clear now, but over time it may not be.
> > > 
> > > 
> > > Can I make a counter-offer of unit tests for this data structure? I think tests would be preferable to asserts, since they'd aid future developers and add no run-time cost (if only for development builds).
> > 
> > 
> > Interesting thought, but I'm not ready to accept it at this time. Here's why:
> > 
> > 1. Unit tests do not make the invariant explicit in the code
> > 2. I suspect that adding an assert can be done with minimal overhead. Either one Dictionary-of-Sets probe in each direction after modification, or a verify pass at the end of the run. I'd feel better going that route.
> > 3. A pitfall of unit tests--which I've fallen into myself from time to time--is that by testing specific inputs, a unit test can miss cases an assert would catch.
> 
> Okay, I added an assert to check that removeOccurrences(of:) removes all occurrences from self.dictionary. I'm having trouble translating your direction from (2) into code that makes sense. I thought about the following:
> 
> ```
>   public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
>     self.reverseIndex[v, default: []].insert(key)
>     let inserted = self.dictionary[key, default: []].insert(v).inserted
>     assert(self.reverseIndex[v]?.contains(key) == true)
>     assert(self.dictionary[key]?.contains(v) == true)
>     return inserted
>   }
> ```
> 
> But it seems silly to assert that the inserts or removals worked as expected. Let me know what you think.

Thank you. Sorry about being unclear. I'll take a look and see if I can clarify.

davidungar commented:
> > > > > Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency? The code is fairly short and clear now, but over time it may not be.
> > > > 
> > > > 
> > > > Can I make a counter-offer of unit tests for this data structure? I think tests would be preferable to asserts, since they'd aid future developers and add no run-time cost (if only for development builds).
> > > 
> > > 
> > > Interesting thought, but I'm not ready to accept it at this time. Here's why:
> > > 
> > > 1. Unit tests do not make the invariant explicit in the code
> > > 2. I suspect that adding an assert can be done with minimal overhead. Either one Dictionary-of-Sets probe in each direction after modification, or a verify pass at the end of the run. I'd feel better going that route.
> > > 3. A pitfall of unit tests--which I've fallen into myself from time to time--is that by testing specific inputs, a unit test can miss cases an assert would catch.
> > 
> > 
> > Okay, I added an assert to check that removeOccurrences(of:) removes all occurrences from self.dictionary. I'm having trouble translating your direction from (2) into code that makes sense. I thought about the following:
> > ```
> >   public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
> >     self.reverseIndex[v, default: []].insert(key)
> >     let inserted = self.dictionary[key, default: []].insert(v).inserted
> >     assert(self.reverseIndex[v]?.contains(key) == true)
> >     assert(self.dictionary[key]?.contains(v) == true)
> >     return inserted
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But it seems silly to assert that the inserts or removals worked as expected. Let me know what you think.
> 
> Thank you. Sorry about being unclear. I'll take a look and see if I can clarify.

OK, after spending some time staring at this, here are my thoughts:
First, I would suggest renaming the member `dictionary` to `forwardDictionary` for the sake of clarity.
Then, instead of the assertion you added, I think it would suffice to just check consistency after every mutation.
That means adding an assertion to the end of `insertValue` and another to the end of `removeValue`.
Here's a sketch, I haven't actually tried these:: ```
 public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
    let reverseInserted = self.reverseIndex[v, default: []].insert(key)
    let forwardInserted = self.dictionary[key, default: []].insert(v).inserted
    assert(forwardInserted = reverseInstered)
    return forwardInserted
  }
```
Removal is a bit trickier: ```
  public mutating func removeValue(_ v: Value, forKey key: Key) -> Value? {
    let reverseOldKey = self.reverseIndex[v]?.remove(key)
    let forwardOldValue = self.dictionary[key, default: []].remove(v)
    assert(isConsistentAfterRemoval(key: key, value: v, oldKey: reverseOldKey, oldValue: reverseOldValue))
    return forwardOldValue
  }
  
private func isConsistentAfterRemoval(key: Key, value: Value, oldKey: Key?, oldValue: Value?))
    switch (oldKey, oldValue) {
      case (nil, nil): return true
      case let (oldKey?, oldValue?) where oldKey == key, oldValue == value: return true
      default: return false
    }
  }
```
Please take a look and see what you think. Thanks!

kbongort commented:
> > > > > > Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency? The code is fairly short and clear now, but over time it may not be.
> > > > > 
> > > > > 
> > > > > Can I make a counter-offer of unit tests for this data structure? I think tests would be preferable to asserts, since they'd aid future developers and add no run-time cost (if only for development builds).
> > > > 
> > > > 
> > > > Interesting thought, but I'm not ready to accept it at this time. Here's why:
> > > > 
> > > > 1. Unit tests do not make the invariant explicit in the code
> > > > 2. I suspect that adding an assert can be done with minimal overhead. Either one Dictionary-of-Sets probe in each direction after modification, or a verify pass at the end of the run. I'd feel better going that route.
> > > > 3. A pitfall of unit tests--which I've fallen into myself from time to time--is that by testing specific inputs, a unit test can miss cases an assert would catch.
> > > 
> > > 
> > > Okay, I added an assert to check that removeOccurrences(of:) removes all occurrences from self.dictionary. I'm having trouble translating your direction from (2) into code that makes sense. I thought about the following:
> > > ```
> > >   public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
> > >     self.reverseIndex[v, default: []].insert(key)
> > >     let inserted = self.dictionary[key, default: []].insert(v).inserted
> > >     assert(self.reverseIndex[v]?.contains(key) == true)
> > >     assert(self.dictionary[key]?.contains(v) == true)
> > >     return inserted
> > >   }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > But it seems silly to assert that the inserts or removals worked as expected. Let me know what you think.
> > 
> > 
> > Thank you. Sorry about being unclear. I'll take a look and see if I can clarify.
> 
> OK, after spending some time staring at this, here are my thoughts: First, I would suggest renaming the member `dictionary` to `forwardDictionary` for the sake of clarity. Then, instead of the assertion you added, I think it would suffice to just check consistency after every mutation. That means adding an assertion to the end of `insertValue` and another to the end of `removeValue`. Here's a sketch, I haven't actually tried these:
> 
> ```
>  public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
>     let reverseInserted = self.reverseIndex[v, default: []].insert(key)
>     let forwardInserted = self.dictionary[key, default: []].insert(v).inserted
>     assert(forwardInserted = reverseInstered)
>     return forwardInserted
>   }
> ```
> 
> Removal is a bit trickier:
> 
> ```
>   public mutating func removeValue(_ v: Value, forKey key: Key) -> Value? {
>     let reverseOldKey = self.reverseIndex[v]?.remove(key)
>     let forwardOldValue = self.dictionary[key, default: []].remove(v)
>     assert(isConsistentAfterRemoval(key: key, value: v, oldKey: reverseOldKey, oldValue: reverseOldValue))
>     return forwardOldValue
>   }
>   
> private func isConsistentAfterRemoval(key: Key, value: Value, oldKey: Key?, oldValue: Value?))
>     switch (oldKey, oldValue) {
>       case (nil, nil): return true
>       case let (oldKey?, oldValue?) where oldKey == key, oldValue == value: return true
>       default: return false
>     }
>   }
> ```
> 
> Please take a look and see what you think. Thanks!

Well, I'm still not sold on having assertions at all, but if we must: I don't think removing the assertion that I added is a good idea, because it covers the most complex mutation that can happen. For the post-removal assertion, I'm not sure it makes sense to do equality checks between the removed value and the value-to-be-removed. That's essentially testing the Set API. So I think we could pare it down to simply check that both removed values are nil or both are non-nil.

davidungar commented on Sources/SwiftDriver/IncrementalCompilation/Multidictionary.swift:
```diff
@@ -123,8 +130,22 @@
+    if self.reverseIndex[v] != nil {
+      return false
+    }
+    for vals in self.forwardDictionary.values {
```
Not a big deal, but if you would like a more concise version, you could code it this way:
```
return !self.forwardDictionaryValues.contains {$0.contains(v)}
```

kbongort replied:
Thanks for the suggestion, done.

davidungar reviewed (approved):
Looks good!

davidungar commented:
> > > > > > > Nicely done! But, there is now an invariant requiring the `reverseIndex` to be consistent with the forward one. I'm worried about this invariant being completely tacit and unchecked. How about something like adding an assertion on every (forward or backward) probe, checking the consistency? The code is fairly short and clear now, but over time it may not be.
> > > > > > 
> > > > > > 
> > > > > > Can I make a counter-offer of unit tests for this data structure? I think tests would be preferable to asserts, since they'd aid future developers and add no run-time cost (if only for development builds).
> > > > > 
> > > > > 
> > > > > Interesting thought, but I'm not ready to accept it at this time. Here's why:
> > > > > 
> > > > > 1. Unit tests do not make the invariant explicit in the code
> > > > > 2. I suspect that adding an assert can be done with minimal overhead. Either one Dictionary-of-Sets probe in each direction after modification, or a verify pass at the end of the run. I'd feel better going that route.
> > > > > 3. A pitfall of unit tests--which I've fallen into myself from time to time--is that by testing specific inputs, a unit test can miss cases an assert would catch.
> > > > 
> > > > 
> > > > Okay, I added an assert to check that removeOccurrences(of:) removes all occurrences from self.dictionary. I'm having trouble translating your direction from (2) into code that makes sense. I thought about the following:
> > > > ```
> > > >   public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
> > > >     self.reverseIndex[v, default: []].insert(key)
> > > >     let inserted = self.dictionary[key, default: []].insert(v).inserted
> > > >     assert(self.reverseIndex[v]?.contains(key) == true)
> > > >     assert(self.dictionary[key]?.contains(v) == true)
> > > >     return inserted
> > > >   }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > But it seems silly to assert that the inserts or removals worked as expected. Let me know what you think.
> > > 
> > > 
> > > Thank you. Sorry about being unclear. I'll take a look and see if I can clarify.
> > 
> > 
> > OK, after spending some time staring at this, here are my thoughts: First, I would suggest renaming the member `dictionary` to `forwardDictionary` for the sake of clarity. Then, instead of the assertion you added, I think it would suffice to just check consistency after every mutation. That means adding an assertion to the end of `insertValue` and another to the end of `removeValue`. Here's a sketch, I haven't actually tried these:
> > ```
> >  public mutating func insertValue(_ v: Value, forKey key: Key) -> Bool {
> >     let reverseInserted = self.reverseIndex[v, default: []].insert(key)
> >     let forwardInserted = self.dictionary[key, default: []].insert(v).inserted
> >     assert(forwardInserted = reverseInstered)
> >     return forwardInserted
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Removal is a bit trickier:
> > ```
> >   public mutating func removeValue(_ v: Value, forKey key: Key) -> Value? {
> >     let reverseOldKey = self.reverseIndex[v]?.remove(key)
> >     let forwardOldValue = self.dictionary[key, default: []].remove(v)
> >     assert(isConsistentAfterRemoval(key: key, value: v, oldKey: reverseOldKey, oldValue: reverseOldValue))
> >     return forwardOldValue
> >   }
> >   
> > private func isConsistentAfterRemoval(key: Key, value: Value, oldKey: Key?, oldValue: Value?))
> >     switch (oldKey, oldValue) {
> >       case (nil, nil): return true
> >       case let (oldKey?, oldValue?) where oldKey == key, oldValue == value: return true
> >       default: return false
> >     }
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please take a look and see what you think. Thanks!
> 
> Thank you for illustrating your suggestion!
> 
> While I'm still not sold on having assertions at all, if we must: I don't think removing the assertion that I added is a good idea, because it covers the most complex mutation that can happen. For the post-removal assertion, I'm not sure it makes sense to do equality checks between the removed value and the value-to-be-removed. That's essentially testing the Set API. So I think we could pare it down to simply check that both removed values are nil or both are non-nil.
> 
> I've updated the PR accordingly.

Yes, I agree with both leaving the longer assertion in and the nil/non-nil check on the other assertion. 
Thanks for your patience!

kbongort commented:
Thank you for the review David! This is my first PR to the project so I'm not familiar with the process – please let me know if there is any further action I need to take.

cltnschlosser commented:
@swift-ci please test

kbongort commented:
<friendly bump> Can we merge this now?

davidungar commented:
> Can we merge this now?

Fine by me. Meant to imply that by my approval. Go for it!

kbongort commented:
I don't think I have the appropriate level of permission to do it myself.

kbongort commented:
(nudge) I don't seem to be able to merge this myself, so if someone else could do so that would be great – thanks!

davidungar commented:
Done!

kbongort commented:
Thank you! 🥳
